# A high-resolution atlas of intergenic transcription in nearly 29,000 normal and cancer samples.

![DOI](https://zenodo.org/badge/DOI/10.5281/zenodo.7740073.svg)](https://doi.org/10.5281/zenodo.7740073)

This is the code repository for the research paper [A high-resolution atlas of intergenic transcription in nearly 29,000 normal and cancer samples](https://www.biorxiv.org/) by Pierre de Langen et al. 

## Table of Contents

- [Background](#background)
- [Installation](#installation)
- [Dependencies](#dependencies)
- [Usage](#usage)
- [Data](#data)
- [License](#license)
- [Acknowledgements](#acknowledgements)
- [References](#references)

## Background

This study present an atlas of active intergenic regions that connects genomic, transcriptomic, and clinical data across normal and cancer samples. The atlas contains over 180,000 RNA Polymerase II bound intergenic regions in more than 900 samples, revealing extensive activity in non-coding regions. By applying a meta-clustering methodology, the authors identified novel per-cancer and pan-cancer intergenic transcriptional markers showing genomic and clinically relevant characteristics, demonstrating the effectiveness of a simple yet effective integration uncovering intergenic transcriptional activity in both normal and cancer tissues.

## Installation

The installation instructions for the project can be found in the `INSTALL.md` file in the repository.

1. Clone the repo
   ```sh
   git clone https://github.com/benoitballester/Pol2Atlas.git

## Dependencies

- Python (version X.X)
- R (version X.X)
- Other software dependencies...

## Usage

The usage instructions for the project can be found in the `USAGE.md` file in the repository[@remap2022].

## Data
![DOI](https://zenodo.org/badge/DOI/10.5281/zenodo.7740073.svg)](https://doi.org/10.5281/zenodo.7740073)

The data used in this project is publicly available at Zenodo and can be found at the following DOI:

https://doi.org/10.5281/zenodo.7740073 

The data includes X, Y and Z  provided in a standardized format for easy analysis.

## License

This project is licensed under the GPL-3.0 license. See the `LICENSE.md` file in the repository for more details.

## Acknowledgements

This work was supported with ; PhD Fellowship to P.D.L. from the French Ministry of Higher Education and Research (MESR); PhD Fellowship to F.H. from the Provence-Alpes-Côte d’Azur Regional Council (Région SUD); Institut National de la Santé et de la Recherche Médicale (INSERM); The Core Cluster of the Institut Français de Bioinformatique (IFB) (ANR-11-INBS-0013) and the Centre de Calcul Intensif d’Aix-Marseille for granting access to its high performance computing resources. The results shown here are based upon data generated by the TCGA Research Network: https://www.cancer.gov/tcga, the GTEx project, the ENCODE Consortium, the ENCODE production laboratories, and independent laboratories who followed the Open Science principles and submitted raw ChIP-seq data into repositories.

## References

\bibliography{references.bib}
